allennlp/nn/beam_search.py

```diff
@@ -524,6 +527,176 @@ def score(
         return average_log_probs
 
 
+class Constraint(Registrable):
+    """
+    An abstract class that can be used to enforce constraints on the output predictions
+    by manipulate the class log probabilities during beam search.
+
+    A `Constraint` just has three methods which need to be implemented: `init_state()`,
+    `apply()` and `update_state()`.
+
+    `init_state()` takes one argument:
+
+    - the batch size, an int
+
+    It returns a constraint state, which is a nested list of dictionaries, with any state needed for subsequent
+    calls to `apply()` and `update_state()`. The length of the outer list should be equal to `batch_size`.
+    Each inner list should be of length 1.
+
+    `apply()` takes two arguments:
+
+    - the constraint state, which is a nested list of dictionaries. The length of the outer list is `batch_size`
+    and the length of each inner list is `beam_size` except on the first time `apply()` is called when it is 1.
+    - `class_log_probabilities`, a tensor of shape `(batch_size, beam_size, num_classes)` that contains the
+    log probabilities for the classes during search. The first time `apply()` is called, `beam_size = 1`.
+
+    The `apply()` method should manipulate the `class_log_probabilities` in place to enforce the constraint
```

Comment:
I think it's more natural for the `apply()` method to return new `class_log_probabilities`. It's a little easier to reason about.